File: src/prototype/web/src/state/posture_break.ts
// PURE — posture_break update per docs/design/state_machines_v1.md §3.
//
// Model: 2D vector (lateral, sagittal). Updated each sim step by combining:
//   1. exponential decay toward origin (τ = 800ms)
//   2. attacker hip input contribution
//   3. GRIPPED-hand "pull" contribution
//   4. defender recovery subtraction
//
// Coefficient values (K*) are placeholders; the design doc explicitly
// defers numeric tuning to post-M1. They live in one place so the
// playtest-driven rebalance is a single edit.

import type { HipIntent } from "../input/intent.js";
import type { Vec2 } from "./game_state.js";

export type PostureBreakConfig = Readonly<{
  decayTauMs: number;       // §3.3 time constant for self-decay
  kHipImpact: number;       // per-second magnitude contributed by 1.0 hip input
  kGripImpact: number;      // per-second magnitude contributed by a 1.0-strength grip
  kRecovery: number;        // per-second magnitude subtracted by 1.0 recovery input
  maxMagnitude: number;     // final clamp; 1.0 is the normative upper bound
}>;

export const DEFAULT_POSTURE_CONFIG: PostureBreakConfig = Object.freeze({
  decayTauMs: 800,
  kHipImpact: 0.9,
  kGripImpact: 0.6,
  kRecovery: 1.2,
  maxMagnitude: 1,
});

// What the update step needs to know from the world. Kept narrow so the
// caller can assemble it without pulling the whole GameState.
export type PostureBreakInputs = Readonly<{
  dtMs: number;              // game_dt (already time-scaled)
  // Attacker (bottom) contributions — drive posture break of the TOP actor.
  attackerHip: HipIntent;
  // Per-hand active pull: non-zero only when that hand is GRIPPED. The
  // direction encodes where the pull is dragging the opponent's upper body
  // in (lateral, sagittal) space. Caller is responsible for mapping
  // GripZone + grip_strength into a direction vector.
  gripPulls: ReadonlyArray<Vec2>;
  // Defender (top) recovery input: unit vector aimed at the origin scaled
  // by the recovery gain (e.g. from a BTN_BASE hold on the top side).
  // Stage 1 currently always passes ZERO_VEC2 here because top-side input
  // is not wired yet; preserved in the signature so the defender commit
  // is a single call-site change.
  defenderRecovery: Vec2;
}>;

export function updatePostureBreak(
  prev: Vec2,
  inputs: PostureBreakInputs,
  cfg: PostureBreakConfig = DEFAULT_POSTURE_CONFIG,
): Vec2 {
  const dtSec = inputs.dtMs / 1000;

  // 1. Exponential decay. Over one dt the state retains exp(-dt/τ).
  const decay = Math.exp(-inputs.dtMs / cfg.decayTauMs);
  let x = prev.x * decay;
  let y = prev.y * decay;

  // 2. Attacker hip contribution. hip_lateral maps to lateral push, and
  // hip_push maps to sagittal — a forward hip thrust (+) tips the opponent
  // forward. The mapping is a design choice that could be revisited, but
  // it's the most direct reading of §B.1 sign conventions.
  x += inputs.attackerHip.hip_lateral * cfg.kHipImpact * dtSec;
  y += inputs.attackerHip.hip_push * cfg.kHipImpact * dtSec;

  // 3. GRIPPED-hand pulls. Each entry is a direction vector with magnitude
  // proportional to (grip_strength · pull_efficacy); the caller already
  // baked those in. Summed in because "more grips = more pulling power"
  // is the desired BJJ feel.
  for (const p of inputs.gripPulls) {
    x += p.x * cfg.kGripImpact * dtSec;
    y += p.y * cfg.kGripImpact * dtSec;
  }

  // 4. Defender recovery subtracts in the direction of the recovery input.
  x -= inputs.defenderRecovery.x * cfg.kRecovery * dtSec;
  y -= inputs.defenderRecovery.y * cfg.kRecovery * dtSec;

  // Clamp magnitude so the vector stays inside the unit disc. Over-shoot
  // is physically meaningless (the opponent can't be "more than broken")
  // and would make §3.4's quantization mapping to paper-proto buckets 0–4
  // leak beyond 4.
  const mag = Math.hypot(x, y);
  if (mag > cfg.maxMagnitude) {
    const s = cfg.maxMagnitude / mag;
    x *= s;
    y *= s;
  }

  return Object.freeze({ x, y });
}

// §3.2 — derived query: is the overall break magnitude above a threshold?
// Useful for judgment-window conditions that reference "‖posture_break‖".
export function breakMagnitude(v: Vec2): number {
  return Math.hypot(v.x, v.y);
}

// §3.4 — paper-proto bucket (0..4). Primarily a HUD / debugging helper.
export function breakBucket(v: Vec2): 0 | 1 | 2 | 3 | 4 {
  const m = breakMagnitude(v);
  if (m < 0.1) return 0;
  if (m < 0.3) return 1;
  if (m < 0.5) return 2;
  if (m < 0.75) return 3;
  return 4;
}

// Encode a per-hand grip pull for §3.3 bullet 3.
// GripZone → unit direction in (lateral, sagittal) space. These are
// intentionally coarse; the real authority for grip→break mapping is
// playtest data, not this table.
import type { GripZone } from "../input/intent.js";

const ZONE_PULL_DIR: Readonly<Record<GripZone, Vec2>> = Object.freeze({
  // Sleeves pulled toward the attacker → forward-break on that side.
  SLEEVE_L: Object.freeze({ x: -0.5, y:  0.87 }), // forward + left
  SLEEVE_R: Object.freeze({ x:  0.5, y:  0.87 }),
  // Collar pulls are strong sagittal break (dumps the opponent forward).
  COLLAR_L: Object.freeze({ x: -0.3, y:  0.95 }),
  COLLAR_R: Object.freeze({ x:  0.3, y:  0.95 }),
  // Wrists pulled laterally drag the opponent sideways.
  WRIST_L:  Object.freeze({ x: -1,   y:  0 }),
  WRIST_R:  Object.freeze({ x:  1,   y:  0 }),
  // Belt pull drags the opponent forward and down.
  BELT:     Object.freeze({ x:  0,   y:  0.9 }),
  // POSTURE_BREAK pulls straight forward with full sagittal weight.
  POSTURE_BREAK: Object.freeze({ x: 0, y: 1 }),
});

export function gripPullVector(zone: GripZone, strength: number): Vec2 {
  const dir = ZONE_PULL_DIR[zone];
  return Object.freeze({ x: dir.x * strength, y: dir.y * strength });
}

File: src/prototype/unity/Assets/BJJSimulator/Runtime/State/PostureBreak.cs
// Ported 1:1 from src/prototype/web/src/state/posture_break.ts.
// PURE — posture_break update per docs/design/state_machines_v1.md §3.
//
// Model: 2D vector (lateral, sagittal). Updated each sim step by combining:
//   1. exponential decay toward origin (τ = 800ms)
//   2. attacker hip input contribution
//   3. GRIPPED-hand "pull" contribution
//   4. defender recovery subtraction

namespace BJJSimulator
{
    // -------------------------------------------------------------------------
    // Config (§3.3 coefficient table — numeric values are design placeholders)
    // -------------------------------------------------------------------------

    public struct PostureBreakConfig
    {
        public float DecayTauMs;           // §3.3 time constant for self-decay
        public float KHipImpact;           // per-second magnitude from 1.0 hip input
        public float KGripImpact;          // per-second magnitude from 1.0-strength grip
        public float KRecovery;            // per-second magnitude subtracted by recovery
        public float MaxMagnitude;         // final clamp; 1.0 is the normative upper bound

        public static readonly PostureBreakConfig Default = new PostureBreakConfig
        {
            DecayTauMs  = 800f,
            KHipImpact  = 0.9f,
            KGripImpact = 0.6f,
            KRecovery   = 1.2f,
            MaxMagnitude = 1f,
        };
    }

    // -------------------------------------------------------------------------
    // Inputs
    // -------------------------------------------------------------------------

    public struct PostureBreakInputs
    {
        public float     DtMs;             // game_dt (already time-scaled), ms
        public HipIntent AttackerHip;
        // Per-hand active pull vectors (non-zero only when GRIPPED).
        // Caller assembles via PostureBreakOps.GripPullVector.
        public Vec2[]    GripPulls;        // may be null or empty
        // Defender recovery input (unit vector × recovery gain). Caller passes
        // Vec2.Zero until defender input is wired.
        public Vec2      DefenderRecovery;
    }

    // -------------------------------------------------------------------------
    // Pure operations
    // -------------------------------------------------------------------------

    public static class PostureBreakOps
    {
        public static Vec2 Update(
            Vec2 prev,
            PostureBreakInputs inp,
            PostureBreakConfig cfg = default)
        {
            if (cfg.DecayTauMs == 0f) cfg = PostureBreakConfig.Default;

            float dtSec = inp.DtMs / 1000f;

            // 1. Exponential decay. Over one dt the state retains exp(-dt/τ).
            float decay = (float)System.Math.Exp(-inp.DtMs / cfg.DecayTauMs);
            float x = prev.X * decay;
            float y = prev.Y * decay;

            // 2. Attacker hip contribution.
            x += inp.AttackerHip.HipLateral * cfg.KHipImpact * dtSec;
            y += inp.AttackerHip.HipPush    * cfg.KHipImpact * dtSec;

            // 3. GRIPPED-hand pulls.
            if (inp.GripPulls != null)
            {
                foreach (var p in inp.GripPulls)
                {
                    x += p.X * cfg.KGripImpact * dtSec;
                    y += p.Y * cfg.KGripImpact * dtSec;
                }
            }

            // 4. Defender recovery subtracts in the direction of the recovery input.
            x -= inp.DefenderRecovery.X * cfg.KRecovery * dtSec;
            y -= inp.DefenderRecovery.Y * cfg.KRecovery * dtSec;

            // Clamp magnitude to the unit disc.
            float mag = (float)System.Math.Sqrt(x * x + y * y);
            if (mag > cfg.MaxMagnitude)
            {
                float s = cfg.MaxMagnitude / mag;
                x *= s;
                y *= s;
            }

            return new Vec2(x, y);
        }

        // §3.2 — derived query: overall break magnitude.
        public static float BreakMagnitude(Vec2 v) =>
            (float)System.Math.Sqrt(v.X * v.X + v.Y * v.Y);

        // §3.4 — paper-proto bucket (0..4). HUD / debug helper.
        public static int BreakBucket(Vec2 v)
        {
            float m = BreakMagnitude(v);
            if (m < 0.1f) return 0;
            if (m < 0.3f) return 1;
            if (m < 0.5f) return 2;
            if (m < 0.75f) return 3;
            return 4;
        }

        // Encode a per-hand grip pull for §3.3 bullet 3.
        // GripZone → unit direction × strength in (lateral, sagittal) space.
        public static Vec2 GripPullVector(GripZone zone, float strength)
        {
            Vec2 dir = ZonePullDir(zone);
            return new Vec2(dir.X * strength, dir.Y * strength);
        }

        static Vec2 ZonePullDir(GripZone zone)
        {
            switch (zone)
            {
                case GripZone.SleeveL:      return new Vec2(-0.5f,  0.87f);
                case GripZone.SleeveR:      return new Vec2( 0.5f,  0.87f);
                case GripZone.CollarL:      return new Vec2(-0.3f,  0.95f);
                case GripZone.CollarR:      return new Vec2( 0.3f,  0.95f);
                case GripZone.WristL:       return new Vec2(-1f,    0f);
                case GripZone.WristR:       return new Vec2( 1f,    0f);
                case GripZone.Belt:         return new Vec2( 0f,    0.9f);
                case GripZone.PostureBreak: return new Vec2( 0f,    1f);
                default:                    return Vec2.Zero;
            }
        }
    }
}
